Add ASSUMEs for UTF-8 byte lengths

khwilliamson opened the pull request:
The maximum number of bytes in a Perl extended UTF-8 character is 13 on ASCII platforms; 14 on EBCDIC.  Yet the variable that returns that number is a Size_t.  By adding these clues to these inline functions, the compiler may be able to do some optimizations.

This isn't done here on another inline function, utf8_to_uv_msgs(). That is because it currently returns the call of a non-inline function, so the ASSUME would be unreachable code.  I don't know if that actually matters.  Or that function's boolean result could be stored in a temporary the ASSUME done, and then utf8_to_uv_msgs() would return the temporary's value.

Opinions welcome


* This set of changes does not require a perldelta entry.

khwilliamson committed:
Add ASSUMEs for UTF-8 byte lengths

The maximum number of bytes in a Perl extended UTF-8 character is 13 on
ASCII platforms; 14 on EBCDIC.  Yet the variable that returns that
number is a Size_t in the cases changed by this commit.  By adding these
ASSUMES to these functions, the compiler may be able to do some
optimizations.

I looked through the code base, and found no other instances where such
a small value could be stored in a fully wide variable.

With link time optimization, an ASSUME may be helpful even in non-inline
functions.

tonycoz reviewed (approved):
I do wonder if assuming `expectlen >= 1` (etc) is useful.

khwilliamson commented:
@tonycoz Any ideas on the second paragraph of https://github.com/Perl/perl5/pull/23680#issue-3381256543

khwilliamson commented:
I added an ASSUME in a place where it would only help link time optimization, and changed as suggested in https://github.com/Perl/perl5/pull/23680#issuecomment-3256966639.

I also audited the code base for other potential spots to change, and found none.
